### PR TITLE
BUGFIX: SD-868 SD-939 Visualization geo chart display Location in the tooltip

### PR DESCRIPTION
**Pull request by dlcd2203**
---

[Check PR owner responsibilities](https://confluence.intgdc.com/display/Development/Code-reviews#Code-reviews-Ownerresponsibilities)

Supported PR commands:

Command | Description
--- | ---
`extended test - examples` | Live examples tests
`extended test - storybook` | Storybook screenshot tests

See more [options](https://confluence.intgdc.com/display/kbhr/How+to+work+with+git+and+Github#HowtoworkwithgitandGithub-Extendedchecks).

# PR Checklist

- [x] Verify code changes ([Checklist](https://confluence.intgdc.com/display/Development/Code-reviews+checklist), [Best practices](https://confluence.intgdc.com/display/Development/Code-reviews+best+practices))
- [x] [Verify pull-request formalities](https://confluence.intgdc.com/display/Development/Code-reviews)
- [x] Change was tested by using [gdc-dev-release](https://confluence.intgdc.com/display/~tomas.vojtasek/Private+NPM) in [gdc-analytical-designer](https://github.com/gooddata/gdc-analytical-designer) and [gdc-dashboards](https://github.com/gooddata/gdc-dashboards) (if applicable)
- [x] Migration guide (for major changes) is mentioned in [CHANGELOG.md](../blob/master/CHANGELOG.md).
- [x] Successful `extended test - examples`
- [x] Successful `extended test - storybook`
- [x] Checked yarn.lock consistency (no dep. duplicities especially Goodstrap)

# Related PRs

- gdc-analytical-designer: https://github.com/gooddata/gdc-analytical-designer/pull/3072
- gdc-dashboards: https://github.com/gooddata/gdc-dashboards/pull/2783

# Related Jira tasks

- SD-868: https://jira.intgdc.com/browse/SD-868
- SD-939: https://jira.intgdc.com/browse/SD-939

### CODE REVIEW

**Comment by dlcd2203** [comment]
extended test - storybook

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-unit-tests-zuul-docker-el7/7205/ : SUCCESS in 5m 36s
- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-oss-check/3197/ : SUCCESS in 3s

**Comment by yenkins** [comment]
Link to Storybook test result:

https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-storybook-zuul-docker/3770/artifact/test-storybook/html-report/index.html

**Comment by gdgate** [comment]
Build succeeded (check-extended pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-oss-check/3198/ : SUCCESS in 3s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-pull-request-dispatcher/3146/ : SUCCESS in 11m 15s

**Comment by dlcd2203** [comment]
extended test - examples

**Comment by yenkins** [comment]
:arrow_right: Started Live Examples deploy to:
https://gdc-live-examples-1573.herokuapp.com

It will be destroyed on merge/close. See [ci-infra](https://github.com/gooddata/ci-infra/blob/master/jenkins/jobs/templates/orchestrators/client-pr-dispatcher.groovy) and [gdc-ci.yaml](https://github.com/gooddata/gooddata-react-components/blob/master/gdc-ci.yaml).

**Comment by gdgate** [comment]
Build succeeded (check-extended pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-oss-check/3199/ : SUCCESS in 4s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-pull-request-dispatcher/3147/ : SUCCESS in 17m 33s

File: src/helpers/MdObjectHelper.ts [hunk 1]
```diff
@@ -84,16 +89,13 @@ export const mdObjectToGeoPushpinBucketProps = (
...
     const location: VisualizationInput.IAttribute =
         locationBucket && (locationBucket.items[0] as VisualizationObject.IVisualizationAttribute);
     const segmentBy: VisualizationInput.IAttribute =
         segmentByBucket && (segmentByBucket.items[0] as VisualizationObject.IVisualizationAttribute);
     const size: VisualizationInput.AttributeOrMeasure = sizeBucket && sizeBucket.items[0];
-    const tooltipText: VisualizationInput.IAttribute =
-        tooltipTextBucket && (tooltipTextBucket.items[0] as VisualizationObject.IVisualizationAttribute);
 
     // we'll use the default internal sorting value which
     // allow us to show the smaller pushpins on top of bigger pushpins
```
Comment (ptquang86) [inline src/helpers/MdObjectHelper.ts:101]: these comment is not correct any more, right?

File: src/helpers/MdObjectHelper.ts [hunk 2]
```diff
@@ -109,17 +111,21 @@ export const mdObjectToGeoPushpinBucketProps = (
...
-        tooltipText,
-    };
+    const { tooltipText: tooltipTextDisplayForm } = config;
+    const configWithTooltipText: IGeoConfig =
+        tooltipTextDisplayForm && typeof tooltipTextDisplayForm === "string" // tooltipText uri was built from properties.control
+            ? {
+                  ...config,
+                  tooltipText: buildAttributeBucket(tooltipTextDisplayForm, TOOLTIP_TEXT),
+              }
+            : config;
```
Comment (ptquang86) [inline src/helpers/MdObjectHelper.ts:121]: How about?
```
const configWithTooltipText: IGeoConfig = { ...config };
if (tooltipTextDisplayForm && typeof tooltipTextDisplayForm === "string") {
     configWithTooltipText.tooltipText = buildAttributeBucket(tooltipTextDisplayForm, TOOLTIP_TEXT);
}
```

File: src/helpers/MdObjectHelper.ts [hunk 3]
```diff
@@ -109,17 +111,21 @@ export const mdObjectToGeoPushpinBucketProps = (
...
 
     return {
-        color,
-        config: configWithTooltipText,
         location,
-        segmentBy,
         size,
+        color,
+        segmentBy,
+        config: configWithTooltipText,
```
Comment (ptquang86) [inline src/helpers/MdObjectHelper.ts:128]: these changes are not necessary :)

File: stories/uri_components/Visualization.tsx [hunk 4]
```diff
@@ -407,6 +408,7 @@ storiesOf("URI components", module)
                     identifier="pushpin-chart-location-size-color-segment-tooltip"
                     config={{
                         mapboxToken: process.env.STORYBOOK_MAPBOX_ACCESS_TOKEN,
+                        tooltipText: attribute("/gdc/md/storybook/obj/24.df").localIdentifier("tooltipText"),
```
Comment (ptquang86) [inline stories/uri_components/Visualization.tsx:411]: this conversation was done [here](https://github.com/gooddata/gooddata-react-components/pull/1573/files?file-filters%5B%5D=.ts&file-filters%5B%5D=.tsx#diff-96359152e9393aae1269513293326debR119), right?

So we only pass the uri to tooltipText is ok
`tooltipText: "/gdc/md/storybook/obj/24.df"`

**Comment by dlcd2203** [comment]
extended test - examples

**Comment by gdgate** [comment]
Merge Failed.

Zuul merger could not merge this change into the base branch. This is most likely caused by merge conflicts. Please rebase the change and upload the rebased version. In case of further errors, contact the zuul administrator.

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-unit-tests-zuul-docker-el7/7239/ : SUCCESS in 5m 20s
- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-oss-check/3235/ : SUCCESS in 4s

**Comment by dlcd2203** [comment]
extended test - examples

**Comment by yenkins** [comment]
:arrow_right: Started Live Examples deploy to:
https://gdc-live-examples-1573.herokuapp.com

It will be destroyed on merge/close. See [ci-infra](https://github.com/gooddata/ci-infra/blob/master/jenkins/jobs/templates/orchestrators/client-pr-dispatcher.groovy) and [gdc-ci.yaml](https://github.com/gooddata/gooddata-react-components/blob/master/gdc-ci.yaml).

**Comment by gdgate** [comment]
Build succeeded (check-extended pipeline).

- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-oss-check/3236/ : SUCCESS in 4s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-pull-request-dispatcher/3166/ : SUCCESS in 17m 17s

File: src/components/uri/Visualization.tsx [hunk 5]
```diff
@@ -388,9 +388,13 @@ export class VisualizationWrapped extends React.Component<
                 );
             }
             case VisualizationTypes.PUSHPIN:
+                const processedVisualizationObject = {
+                    ...mdObject,
+                    content: fillMissingTitles(mdObject.content, locale),
```
Comment (naibinh) [inline src/components/uri/Visualization.tsx:393]: Is this change an improvement?

File: src/helpers/MdObjectHelper.ts (hunk 2 above)
Comment (naibinh) [reply src/helpers/MdObjectHelper.ts:121]: Yes, I agree and want to add some
1. `configWithTooltipText.tooltipText =  attribute(uri).localIdentifier(localIdentifier)` so we don't have to create new 
single-use `buildAttributeBucket `
2. long `tooltipTextDisplayForm && typeof tooltipTextDisplayForm === "string"` should be assigned to meaningful const. And what other types can `tooltipTextDisplayForm ` hold besides `string`?

File: src/components/uri/Visualization.tsx (hunk 5 above)
Comment (dlcd2203) [reply src/components/uri/Visualization.tsx:393]: this change fixes[issue](https://jira.intgdc.com/browse/SD-868?focusedCommentId=1265835&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-1265835)

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-unit-tests-zuul-docker-el7/7240/ : SUCCESS in 5m 36s
- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-oss-check/3237/ : SUCCESS in 3s

File: src/helpers/MdObjectHelper.ts (hunk 1 above)
Comment (dlcd2203) [reply src/helpers/MdObjectHelper.ts]: removed

File: src/helpers/MdObjectHelper.ts (hunk 2 above)
Comment (dlcd2203) [reply src/helpers/MdObjectHelper.ts]: fixed

File: src/helpers/MdObjectHelper.ts (hunk 3 above)
Comment (dlcd2203) [reply src/helpers/MdObjectHelper.ts]: reverted

File: stories/uri_components/Visualization.tsx [hunk 6]
```diff
@@ -394,6 +395,23 @@ storiesOf("URI components", module)
             </div>,
         ),
     )
+    .add("GeoPushpinChart example", () =>
```
Comment (naibinh) [inline stories/uri_components/Visualization.tsx:398]: Please comment out temporarily this test case as it's unstable now.
We'll bring it back once [SD-948](https://jira.intgdc.com/browse/SD-948) is resolved.

Just note, this case just covers SD-939, there should be one case (without config.tooltip) for SD-868

Comment (dlcd2203) [reply stories/uri_components/Visualization.tsx]: SD-868 include 2 case: one is the same with SD-939, one relate with arithmetic measure. So I think this case still have config with Tooltiptext.

**Comment by gdgate** [comment]
Build failed (check pipeline).

- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-unit-tests-zuul-docker-el7/7241/ : FAILURE in 3m 11s
- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-oss-check/3238/ : SUCCESS in 3s

**Comment by ptquang86** [comment]
LGTM

**Comment by gdgate** [comment]
Build failed (check pipeline).

- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-unit-tests-zuul-docker-el7/7242/ : FAILURE in 47s
- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-oss-check/3239/ : SUCCESS in 3s

**Comment by gdgate** [comment]
Build succeeded (check pipeline).

- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-unit-tests-zuul-docker-el7/7243/ : SUCCESS in 3m 32s
- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-oss-check/3240/ : SUCCESS in 3s

**Comment by gdgate** [comment]
Build succeeded (gate pipeline).

- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-storybook-zuul-docker/3797/ : SUCCESS in 12m 26s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-unit-tests-zuul-docker-el7/7244/ : SUCCESS in 7m 53s
- https://checklist.intgdc.com/job/compliance/job/gooddata-react-components-merge-compliance-check/1768/ : SUCCESS in 12s
- https://checklist.intgdc.com/job/client-libs/job/gooddata-react-components-examples-zuul-docker/2595/ : SUCCESS in 17m 34s

**Pull request merged**